### PR TITLE
Fix FSDP hanging for multimodal models

### PR DESCRIPTION
**Pull request by jinsooihm**
## Problem
Consider a mutlimodal model that only runs its vision encoder when an image is present in the input.
Suppose DP = 2 and rank 0 gets an image rank 1 does not.
With FSDP on the vision encoder, rank 0 will issue an all gather, which will be not met because rank 1 does not run its vision encoder. 

This problem exists in the current multimodal models, but the cc12m dataset does not expose it because its samples are always text + image. 

I have only implemented this for Muse Glimmer for now but it can be extended to other models if you agree with the design.

## Proposed solution
1. Run vision encoder with always and if input is None, return an empty tensor, so all gather is met.
2. Add a zero value related to the empty tensor on the embedded tokens to make encoder backward execute, since there is a reduce scatter that needs to be met. 
3. This requires FSDP `set_reduce_scatter_unused_params(True)` which sets grad to 0 for unused parameters. 
The above 3 steps solve the hanging.

## Alternate design
An alternate design could be to wrap the whole vision region + scattering as a separate module `vision_tower` and let `vision_tower.forward(embedded_tokens, pixel_values, ...)` that 
- if `pixel_values` is `None` returns `embedded_tokens` right away 
- run the encoder and scatter and returns `embedded_tokens_with_vision_features`.

We can then wrap `vision_tower` as a whole fsdp region. This can remove the hacky stuff at step 2 in the proposed solution.

I think this could be cleaner, but can change FQN of the vision parts. 

<del>
There is another subtlety coming from optimization. If there are many optimization steps without an image, and we use AdamW, the weight of the vision encoder will keep decaying. To prevent that, we 

1. Record whether the model got an image in the input in `preprocess_inputs`
2. Register a pre optimizer hook that sets gradient of unused parameters to be `None` if no DP rank saw an image in its optimization step, including PP and grad accumulation microbatches. This requires communication and D2H sync between backward and optimizer step. 
4. Then the optimizer does not do optimizer step to those parameters.

This may be a data problem and decay is the intended behavior? Then the solution can be simple by just doing the first 3 steps. 
Please comment, cc. @shuhuayu @tianyu-l 

## Limitations
This makes optimizer step not cudagraphable. 
One way to get around this would be to have to separate optimizers for the decoder and encoder part, and run one conditionally, but that would be more invasive change. 

</del>

### CODE REVIEW

File: torchtitan/models/muse_glimmer/parallelize.py [hunk 1]
```diff
@@ -127,6 +129,10 @@ def parallelize_muse_glimmer(
         dp_mesh_dims=dp_mesh_dims,
         symm_mem_scope=parallelism.fsdp_symm_mem_scope,
     )
+    if has_vision:
+        for module in (model, model.vision_encoder, model.vision_adapter):
+            assert isinstance(module, FSDPModule)
+            module.set_reduce_scatter_unused_params(True, recurse=False)
```
Comment (tianyu-l) [inline torchtitan/models/muse_glimmer/parallelize.py:135]: why don't we always turn this on -- is there a case in torchtitan where it does unecessary reduction?

asking from another angle: what if vision encoder is frozen, do we set this to True?

Comment (fegin) [reply torchtitan/models/muse_glimmer/parallelize.py:135]: `set_reduce_scatter_unused_params` only performs `reduce_scatter` when requires_grad is True. So, it seems that we should always turn on this flag.

Comment (tianyu-l) [reply torchtitan/models/muse_glimmer/parallelize.py:135]: @weifengpy do you know why it's not True by default in FSDP?

Comment (weifengpy) [reply torchtitan/models/muse_glimmer/parallelize.py:135]: this regresses memories. `set_reduce_scatter_unused_params ` is a lazy workaround for non-spmd reduce-scatter. for this case, it's related to some rank didn't trigger RS from vision modules, while other ranks trigger

non-lazy way is to address the root cause of non-spmd vision behavior direclty

Comment (weifengpy) [reply torchtitan/models/muse_glimmer/parallelize.py:135]: numerics are also problematic - some ranks just do zero-valued gradients. but those gradients are averaged with real gradients from other ranks

so address non-spmd vision behavior is more rigor than using this API

Comment (shuhuayu) [reply torchtitan/models/muse_glimmer/parallelize.py:135]: suppose some ranks get vision input, some ranks do not, then i think we should average gradients by treating the current dp rank's vit's gradient as zero?

Comment (weifengpy) [reply torchtitan/models/muse_glimmer/parallelize.py:135]: > suppose some ranks get vision input, some ranks do not, then i think we should average gradients by treating the current dp rank's vit's gradient as zero?

after grad reduction, all real gradients shrinks towards zero a little bit

Comment (shuhuayu) [reply torchtitan/models/muse_glimmer/parallelize.py:135]: >after grad reduction, all real gradients shrinks towards zero a little bit

my understanding is that if we consider the vit + decoder as one single model, and lm loss is defined over all valid tokens, then we should be consistent while scaling the loss/gradient, i.e. using number of valid tokens, whether there are images contributing vit gradients locally is not relevant.

Comment (tianyu-l) [reply torchtitan/models/muse_glimmer/parallelize.py:135]: > numerics are also problematic - some ranks just do zero-valued gradients. but those gradients are averaged with real gradients from other ranks

We discussed this offline. I think this is expected. This sounds more of a data distribution characteristic / problem, rather than infra problem.

I agree that we should keep spmd behavior as much as possible.

Comment (weifengpy) [reply torchtitan/models/muse_glimmer/parallelize.py]: got you. spmd is the main reason then

**Comment by jinsooihm** [comment]
After offline discussion, I changed to feeding a dummy input that goes through the whole encoder forward, that is constructed in `preprocess_inputs` of the model owning the encoder. 

Implemented only for Muse Glimmer yet, and I can proceed for other models if everyone agrees.

**Comment by shuhuayu** [comment]
> After offline discussion, I changed to feeding a dummy input that goes through the whole encoder forward, that is constructed in `preprocess_inputs` of the model owning the encoder.
> 
> Implemented only for Muse Glimmer yet, and I can proceed for other models if everyone agrees.

make sense to me from maintaining static shapes point of view for cudagraphbility.

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/models/qwen3_5/model.py [hunk 2]
```diff
@@ -606,13 +608,27 @@ def _prepare_multimodal_embeds(
         inputs_embeds = (
             self.tok_embeddings(tokens) if self.tok_embeddings is not None else tokens
         )
-
-        if pixel_values is not None and grid_thw is not None:
+        if self.vision_encoder is None:
+            return inputs_embeds
+
+        # TODO: Configure the job-wide modality set from the dataset so
```
Comment (shuhuayu) [inline torchtitan/models/qwen3_5/model.py:614]: currently, for image + text dataset, there will be a normal image fed vit call, and a dummy fed vit call? `video_is_dummy` is still true.

Comment (jinsooihm) [reply torchtitan/models/qwen3_5/model.py:614]: Yes.

I think there could be two ways to solve this.

Add in the model config definition whether this training run will have image, video or image + video inputs. 

Another way could be to make the collator emit interleaved image + video in a single `pixel_values`? It seems like most of the vision encoders do not care if the input is an image or video, since that information is handled by `grid_thw`. Then we can just make the vision encoder run once. Scattering logic may need some change. But this approach would also prevent the fsdp all gather of vision encoder running twice.

Comment (shuhuayu) [reply torchtitan/models/qwen3_5/model.py:614]: does muse glimmer have mixed image-video input? it sounds like if there are mixed images + videos, there are always separate calls of vision encoders. long term cleaner design would be allowing mixing images + videos in one vision encoder call. is this issue specific to qwen 3.5 now?

Comment (jinsooihm) [reply torchtitan/models/qwen3_5/model.py:614]: Currently muse glimmer has video disabled, but enabling it would require just a small change in the encoder code, nothing fundamental. 

The issue is specific for qwen 3.5 now because it is the only one allowing image + video input in the same batch. Other models do not run into this case because
- Kimi 3 and Muse Glimmer: do not support video yet
- Kimi 2.7: special tokens do not distinguish image and video.

Kimi 2.7 todo says
```
        # TODO: support mixed image+video batches. Upstream fix: when
        # image_id == video_id, emit one document-ordered vision stream so the
        # runs stay modality-agnostic and this branch goes away.
```
so I believe this is suggesting the single encoder call direction.

Comment (shuhuayu) [reply torchtitan/models/qwen3_5/model.py]: the kimi 2.7 todo points toward a document-ordered packed media stream, which is consistent with running the vision encoder once. ok, i think mixed image videos input is non-blocking for this pr.

**Review by shuhuayu — APPROVED** [review]

**Comment by shuhuayu** [comment]
@jinsooihm can you help rebase?

**Comment by weifengpy** [comment]
is the final solution trainer specific or can be upstreamed as a fsdp api?

**Comment by jinsooihm** [comment]
@weifengpy The solution we settled for here is trainer specific.

**Comment by weifengpy** [comment]
> @weifengpy The solution we settled for here is trainer specific. 

got you. thanks
